tests/admin/test_routes.py

```diff
@@ -0,0 +1,794 @@
+        self.original_app = self.routes.app
+        self.resolver = self.original_app.url_map.bind('', '/')
+
+        # For convenience, set self.controller to a specific controller
+        # whose routes are being tested.
+        controller_name = getattr(self, 'CONTROLLER_NAME', None)
+        if controller_name:
+            self.controller = getattr(self.manager, controller_name)
+
+            # Make sure there's a controller by this name in the real
+            # CirculationManager.
+            self.real_controller = getattr(
+                self.REAL_CIRCULATION_MANAGER, controller_name
+            )
+        else:
+            self.real_controller = None
+
+        self.routes.app = app
+
+    def assert_authenticated_request_calls(self, url, method, *args, **kwargs):
+        """First verify that an unauthenticated request fails. Then make an
+        authenticated request to `url` and verify the results, as with
+        assert_request_calls
+        """
+        authentication_required = kwargs.pop("authentication_required", True)
+
+        http_method = kwargs.pop('http_method', 'GET')
+        response = self.request(url, http_method)
+        if authentication_required:
+            eq_(401, response.status_code)
+            eq_("authenticated_admin_from_request called without authorizing",
+                response.data)
+        else:
+            eq_(200, response.status_code)
+
+        # Set a variable so that authenticated_admin_from_request
+        # will succeed, and try again.
+        self.manager.admin_sign_in_controller.authenticated = True
+        try:
+            kwargs['http_method'] = http_method
+            if kwargs.get('file_response', None) is not None:
+                self.assert_file_response(url, *args, **kwargs)
+            else:
+                self.assert_request_calls(url, method, *args, **kwargs)
+        finally:
+            # Un-set authentication for the benefit of future
+            # assertions in this test function.
+            self.manager.admin_sign_in_controller.authenticated = False
+
+    def assert_file_response(self, url, *args, **kwargs):
+        http_method = kwargs.pop('http_method', 'GET')
+        response = self.request(url, http_method)
+
+        eq_(response.headers['Content-type'], 'text/csv')
+    
+    def assert_redirect_call(self, url, *args, **kwargs):
+
+        self.manager.admin_sign_in_controller.csrf_token = True
+        self.manager.admin_sign_in_controller.authenticated = True
+        http_method = kwargs.pop('http_method', 'GET')
+        response = self.request(url, http_method)
+
+        # A Flask template string is returned.
+        assert "You are now logged in" in response
+
+        # Either not being authenticated or not having a csrf token fails
+        # but setting both to false.
+        self.manager.admin_sign_in_controller.csrf_token = False
+        self.manager.admin_sign_in_controller.authenticated = False
```

Comment:
What's supposed to happen if one of these is set but not both?

Reply:
Hmm yes, I should add those cases.